### PR TITLE
improve unit-tests for JsonService

### LINKED ISSUE #214: improve unit-tests for JsonService
JsonService.postForm is currently untested, not good for maintenance and we had a bug #213, which is fixed now

### PR DESCRIPTION
**Pull request by pamapa**
closes #214

i am using `"test": "tsc -p test/unit/tsconfig.json && jest --coverage",` to detect missing spots, we have quite some...

### CODE REVIEW

File: test/unit/JsonService.test.ts [hunk 1]
```diff
@@ -30,30 +29,41 @@ describe("JsonService", () => {
...
-            try { await p; } catch {}
+            try {
+                // act
+                await subject.getJson("http://test", "token");
+            } catch {
+                // assert
+                expect(fetchMock).toBeCalledWith("http://test", {
+                    headers: { Authorization: "Bearer token" },
+                    method: "GET"
+                });
```
Comment (kherock) [inline test/unit/JsonService.test.ts:65]: This test will create a false positive if `.getJson` unexpectedly succeeds. Use something like this for testing async errors:

```suggestion
                await expect(subject.getJson("http://test", "token")).rejects.toThrow(Error);
                expect(fetchMock).toBeCalledWith("http://test", {
                    headers: { Authorization: "Bearer token" },
                    method: "GET"
                });

File: test/unit/JsonService.test.ts [hunk 2]
```diff
@@ -74,6 +84,28 @@ describe("JsonService", () => {
...
+            try {
+                // act
+                await subject.getJson("http://test");
+                fail("should not come here");
+            }
+            catch (err) {
+                // assert
+                expect(err).toBeInstanceOf(SyntaxError);
+                expect((err as SyntaxError).message).toContain("Unexpected token a in JSON");
+            }
```
Comment (kherock) [inline test/unit/JsonService.test.ts:106]: Same here, you can reduce this to

```suggestion
            await expect(subject.getJson("http://test")).rejects.toThrow(new SyntaxError("Unexpected token a in JSON"));
```

File: test/unit/JsonService.test.ts [hunk 3]
```diff
@@ -148,5 +182,314 @@ describe("JsonService", () => {
...
+    describe("postForm", () => {
+        it("should return a promise", async () => {
+            // act
+            const p = subject.postForm("http://test", { "a": "b" });
+
+            // assert
+            expect(p).toBeInstanceOf(Promise);
+            // eslint-disable-next-line no-empty
+            try { await p; } catch {}
+        });
```
Comment (kherock) [inline test/unit/JsonService.test.ts:216]: do we need tests like this? They don't add to the line coverage and TypeScript already enforces type safety.

File: test/unit/JsonService.test.ts [hunk 4]
```diff
@@ -148,5 +182,314 @@ describe("JsonService", () => {
...
+                expect(err).toBeInstanceOf(Error);
+                expect((err as Error).message).toContain("url");
+            }
+        });
+
+        it("should make POST request to url", async () => {
+            try {
+                // act
+                await subject.postForm("http://test", { "a": "b" });
+            } catch {
```
Comment (kherock) [inline test/unit/JsonService.test.ts:233]: It seems like a lot of these `try/catch` patterns that only test if the request was attempted should actually be `try/finally`s.

**Comment by pamapa** [comment]
I have now eliminate the error prone try/catch in the two test files with the pattern you suggested. Also the check for promise is now gone.

**Review by kherock — APPROVED** [review]
Awesome these look good now. Also just a heads up, I have some tests improving the coverage of UserManager and PopupWindow that I'll check in once the coverage diffs are up

**Pull request merged**